### PR TITLE
Fix proxy support

### PR DESCRIPTION
**Pull request by Mitko-Kerezov**
* Use `request` npm module which has out-of-the-box support for proxies
* Allow the user to specify proxy protocol - up until now we only had `http`
* Rename `proxy-<action>-command.ts` files to `proxy-<action>.ts`

Ping @rosen-vladimirov @TsvetanMilanov

### CODE REVIEW

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: commands/proxy/proxy-set.ts [hunk 1]
```diff
@@ -22,22 +23,30 @@ export class ProxySetCommand extends ProxyCommandBase {
...
+			if (!isInteractive()) {
+				this.$errors.fail("Console is not interactive - you need to supply all command parameters.");
+			} else {
+				urlString = await this.$prompter.getString("Url", { allowEmpty: false });
+			}
+		}
 
-		const noHostName = !hostname;
+		const urlObj = parse(urlString);
+		let port = urlObj.port && +urlObj.port || HttpProtocolToPort[urlObj.protocol];
```
Comment (rosen-vladimirov) [inline commands/proxy/proxy-set.ts]: is there any way for port to be undefined after this line? In case not, please fix the next line

File: commands/proxy/proxy-set.ts [hunk 2]
```diff
@@ -22,22 +23,30 @@ export class ProxySetCommand extends ProxyCommandBase {
...
 		const noPort = !port || !this.isValidPort(port);
+		const authCredentials = this.getCredentialsFromAuth(urlObj.auth || "");
+		username = username || authCredentials.username;
```
Comment (rosen-vladimirov) [inline commands/proxy/proxy-set.ts]: I believe we should fail in case you specify username in both ways, for example:
`appbuilder proxy set http://username:password@127.0.0.1:8888 username1 password1`
should fail. In case the username and password on both places are the same, we should not fail.

File: http-client.ts [hunk 3]
```diff
@@ -97,65 +95,77 @@ export class HttpClient implements Server.IHttpClient {
...
-			});
-
-			request.on("error", (err: Error) => {
-				this.setResponseResult(promiseActions, timerId, { err });
-			});
+			options.url = options.url || `${options.proto}://${options.host}${options.path}`;
+			options.encoding = null;
+			options.followAllRedirects = true;
+
+			let requestObj = request(options);
```
Comment (rosen-vladimirov) [inline http-client.ts]: `const` maybe

File: http-client.ts [hunk 4]
```diff
@@ -97,65 +95,77 @@ export class HttpClient implements Server.IHttpClient {
...
+
+			requestObj
+				.on("error", (err: Error) => {
+					this.$logger.trace("An error occurred while sending the request:", err);
+					const errorMessageMatch = err.message.match(HttpClient.STATUS_CODE_REGEX);
+					const errorMessageStatusCode = errorMessageMatch && errorMessageMatch[1] && +errorMessageMatch[1];
+					let errorMessage = this.getErrorMessage(errorMessageStatusCode, null);
```
Comment (rosen-vladimirov) [inline http-client.ts]: const

File: http-client.ts [hunk 5]
```diff
@@ -97,65 +95,77 @@ export class HttpClient implements Server.IHttpClient {
 
 			this.$logger.trace("httpRequest: %s", util.inspect(options));
```
Comment (rosen-vladimirov) [inline http-client.ts]: this trace must be immediately before the request - currently it will not print all options that we pass and we may be confused when asking customers for logs.

File: http-client.ts [hunk 6]
```diff
@@ -266,16 +277,19 @@ export class HttpClient implements Server.IHttpClient {
...
+			let credentialsPart = "";
 			const proxyCredentials = await this.$proxyService.getCredentials();
 			if (proxyCredentials && proxyCredentials.username && proxyCredentials.password) {
-				headers["Proxy-Authorization"] = "Basic " + new Buffer(`${proxyCredentials.username}:${proxyCredentials.password}`).toString('base64');
+				credentialsPart = `${proxyCredentials.username}:${proxyCredentials.password}@`;
 			}
 
+			// Note that proto ends with :
+			options.proxy = `${proto}//${credentialsPart}${host}:${port}`;
+			options.rejectUnauthorized = false;
```
Comment (rosen-vladimirov) [inline http-client.ts]: Why is this required? Maybe this should be an option in the proxy configuration.

File: commands/proxy/proxy-set.ts [hunk 7]
```diff
@@ -22,22 +23,30 @@ export class ProxySetCommand extends ProxyCommandBase {
...
+		password = password || authCredentials.password;
 
-		if (!isInteractive() && (noHostName || noPort || this.isPasswordRequired(username, password))) {
+		if (!isInteractive() && (noPort || this.isPasswordRequired(username, password))) {
```
Comment (rosen-vladimirov) [inline commands/proxy/proxy-set.ts]: I know I've written this part, but now I see an incorrect behavior here - In case I've specified port, but it's not valid and the terminal is not interactive, we'll receive error "Console is not interactive - you need to supply all command parameters."
Can you fix this by providing correct message in such case?

File: http-client.ts [hunk 8]
```diff
@@ -97,65 +95,77 @@ export class HttpClient implements Server.IHttpClient {
...
+			options.url = options.url || `${options.proto}://${options.host}${options.path}`;
+			options.encoding = null;
+			options.followAllRedirects = true;
+
+			let requestObj = request(options);
+
+			requestObj
+				.on("error", (err: Error) => {
+					this.$logger.trace("An error occurred while sending the request:", err);
+					const errorMessageMatch = err.message.match(HttpClient.STATUS_CODE_REGEX);
```
Comment (rosen-vladimirov) [inline http-client.ts]: Can you please add comment why we need this regex? You can place [this url](https://github.com/request/tunnel-agent/blob/eb2b1b19e09ee0e6a2b54eb2612755731b7301dc/index.js#L166) in the comment, so in case the response is added to the error object in a later version, we'll know where to search for it.

File: http-client.ts [hunk 9]
```diff
@@ -97,65 +95,77 @@ export class HttpClient implements Server.IHttpClient {
...
+					err.message = errorMessage || err.message;
+					this.setResponseResult(promiseActions, timerId, { err });
+				})
+				.on("response", (response: Server.IRequestResponseData) => {
+					let successful = helpers.isRequestSuccessful(response);
+					if (!successful) {
+						pipeTo = undefined;
+					}
+
+					let responseStream = response;
```
Comment (rosen-vladimirov) [inline http-client.ts]: const

File: http-client.ts [hunk 10]
```diff
@@ -97,65 +95,77 @@ export class HttpClient implements Server.IHttpClient {
...
+				.on("error", (err: Error) => {
+					this.$logger.trace("An error occurred while sending the request:", err);
+					const errorMessageMatch = err.message.match(HttpClient.STATUS_CODE_REGEX);
+					const errorMessageStatusCode = errorMessageMatch && errorMessageMatch[1] && +errorMessageMatch[1];
+					let errorMessage = this.getErrorMessage(errorMessageStatusCode, null);
+					err.message = errorMessage || err.message;
+					this.setResponseResult(promiseActions, timerId, { err });
+				})
+				.on("response", (response: Server.IRequestResponseData) => {
+					let successful = helpers.isRequestSuccessful(response);
```
Comment (rosen-vladimirov) [inline http-client.ts]: I'm not sure if it's possible to come here with something that's not 2.x.x? Have you seen such behavior?

File: http-client.ts [hunk 11]
```diff
@@ -97,65 +95,77 @@ export class HttpClient implements Server.IHttpClient {
...
+					} else {
+						let data: string[] = [];
+
+						responseStream.on("data", (chunk: string) => {
+							data.push(chunk);
+						});
+
+						responseStream.on("end", () => {
+							this.$logger.trace("httpRequest: Done. code = %d", response.statusCode.toString());
+							let responseBody = data.join("");
```
Comment (rosen-vladimirov) [inline http-client.ts]: const

File: commands/proxy/proxy-set.ts (hunk 1 above)
Comment (Mitko-Kerezov) [inline commands/proxy/proxy-set.ts]: Theoretically there is a way for `port` to be undefined if `HttpProtocolToPort[urlObj.protocol]` returns `undefined`. This would happen in the unlikely case where the user enters an url address which makes use of neither `http` nor `https` protocol.

File: http-client.ts (hunk 3 above)
Comment (Mitko-Kerezov) [inline http-client.ts]: Is disputable because calling `.end()` alters the object's state but nonetheless point taken - will change it to `const`

File: http-client.ts (hunk 9 above)
Comment (Mitko-Kerezov) [inline http-client.ts]: `responseStream` is reassigned in case one of the `switch` cases is met on the following lines

File: http-client.ts (hunk 10 above)
Comment (Mitko-Kerezov) [inline http-client.ts]: Theoretically it should be possible (1xx responses for example) but I haven't seen it myself. I prefer to keep the code this way _better safe than sorry_

File: http-client.ts (hunk 6 above)
Comment (Mitko-Kerezov) [inline http-client.ts]: Talked offline and decided to expose a `--` flag to control this behavior

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: commands/proxy/proxy-set.ts [hunk 12]
```diff
@@ -17,25 +18,42 @@ export class ProxySetCommand extends ProxyCommandBase {
...
 		}
-
-		if (noHostName) {
-			hostname = await this.$prompter.getString("Hostname", { allowEmpty: false });
+		username = username || authCredentials.username;
+		password = password || authCredentials.password;
+
+		if (!isInteractive()) {
+			if (noPort) {
+				this.$errors.failWithoutHelp(`The port you have specified (${port || "none"}) is not valid.`);
```
Comment (rosen-vladimirov) [inline commands/proxy/proxy-set.ts]: `The port you have specified none is not valid`

This seems like a strange error message.

Comment (Mitko-Kerezov) [inline commands/proxy/proxy-set.ts]: It'll actually say
`The port you have specified (none) is not valid`

Moreover it'll be very difficult to get to this error message 😄
